service/header/verify.go

```diff
@@ -0,0 +1,52 @@
+package header
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+)
+
+func VerifyAdjacent(trusted, untrusted *ExtendedHeader) error {
```

Comment:
Doesn't this logic exist in tendermint and we can use it as a library instead?

Reply:
Yeah and I import their code initially, but I decided to reimplement it, as they have a trust period for headers and we don't need them AFAIU, because we need to be able to sync anytime. Also, it looks cleaner when we use our data structure and it us just less code.

One interesting thing though. They also have `VerifyUnadjasent`, where you can verify a new header by a validator set from the past. If there is some overlap then the Header is secure. This is also proof of why backward syncing will work.

Reply:
>  as they have a trust period for headers and we don't need them AFAIU, 

I don't understand why in our case we would not need a trust period. Same as in the tendermint case a trusted header expires after the trust period.

> If there is some overlap then the Header is secure. This is also proof of why backward syncing will work.

True. But the use-case is slightly different because there the headers in between the trusted height and the new height are skipped because of the overlap of the validators. This is thoroughly analyzed and well within the security assumptions of tendermint. We should do the same if we want to implement backward syncing (cc @musalbas @adlerjohn)

Reply:
> I don't understand why in our case we would not need a trust period. Same as in the tendermint case a trusted header expires after the trust period.

What is trust period then and why do __we__ need? Note that the Store is for both light/full nodes and when you sync from scratch you don't have a trust period for headers you request. Also, we verify headers cryptographically, so if it's valid by the PoS rules, why would we need a trust period?

Reply:
> What is trust period then and why do **we** need it?

You can go offline as a light client for any duration < trust period without having to be subjectively reinitialized. It's the same use-case as for tendermint.

Reply:
> Also, we verify headers cryptographically, so if it's valid by the PoS rules, why would we need a trust period for them?

https://blog.ethereum.org/2014/11/25/proof-stake-learned-love-weak-subjectivity/

Reply:
@liamsi, Ok, I see why I was wrong. Thank you for the explanation!

Reply:
I created an issue to track the resolution of long-range attacks https://github.com/celestiaorg/celestia-node/issues/205(currently empty). As discussed in sync, let's not fix the issue before devnet. Also, importing Tendermint implementation of `VerifyAdjacent` would not just help as we should also implement additional logic for unboding period checking.